### ISSUE TITLE
Max registration ID's

### ISSUE THREAD

**Issue by leearmstrong** [ToothlessGear/node-gcm#42]
Will node-gcm fail if I send it more than 1000 Registrations ID's in a single message as that is all Google support?  Do I have to split it up myself before hand?

**Comment by eladnava** [comment]
If anyone else needs to support more than 1,000 devices, you can easily split the tokens yourself like this:

    // Max devices per request    
    var batchLimit = 1000;
    
    // Batches will be added to this array
    var tokenBatches = [];
    
    // Traverse tokens and split them up into batches of 1,000 devices each  
    for ( var start = 0; start < tokens.length; start += batchLimit )
    {
        // Get next 1,000 tokens
        var slicedTokens = tokens.splice(start, start + batchLimit);
        
        // Add to batches array
        tokenBatches.push(slicedTokens);
    }
    
    // You can now send a push to each batch of devices, in parallel, using the caolan/async library
    async.each( batches, function( batch, callback )
    {
        // Assuming you already set up the sender and message
        sender.send(message, { registrationIds: batch }, function (err, result)
        {
            // Push failed?
            if (err)
            {
                // Stops executing other batches
                return callback(err);
            }

            // Done with batch
            callback();
        }
    },
    function( err )
    {
        // Log the error to console
        if ( err )
        {
            console.log(err);
        }
    });

**Comment by hypesystem** [comment]
I wrote parallel-batch which does pretty much that: https://www.npmjs.com/package/parallel-batch

**Comment by eladnava** [comment]
Nice package! Would greatly simplify the code I wrote.

**Comment by eladnava** [comment]
@hypesystem maybe it would be a good idea to integrate your parallel-batch library into node-gcm, so that batching will be performed automagically?

**Comment by hypesystem** [comment]
I agree, that was the original intention with parallel-batch. As it turns out, though, it's easier said than done.

Specifically we want to return errors correctly (as if no batching happened; so if only some of the batches fail, some of the messages will still (possibly) be sent); and we want to handle retries the way the user expects (which is also hard).

**Comment by eladnava** [comment]
I think it's doable -- in case one of the batches failed, we'll retry it until we run out of `tries`. We just have to make sure that GCM doesn't deliver the push notification to some devices in the batch while erroring out -- that would cause some serious spamming. But in any case, it could be happening today with < 1,000 devices in `sender.send`.

**Comment by hypesystem** [comment]
I would love to see you give it a try. First of all, though, I think we need a new issue to discuss this -- feel free to create it.

I will try to gather my thoughts on what behaviour I think we would want, and why, exactly, it is tricky :-)
